src/base/path.h

```diff
@@ -0,0 +1,114 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QMetaType>
+#include <QString>
+
+class Path final
+{
+public:
+    enum class Format
+    {
+        Uniform,
+        Native
+    };
+
+    Path() = default;
+    Path(const Path &other) = default;
+    Path(Path &&other) = default;
+
+    explicit Path(const QString &pathStr);
+    explicit Path(const char pathStr[]);
+    explicit Path(const std::string &pathStr);
+
+    bool isValid() const;
+    bool isEmpty() const;
+    bool isAbsolute() const;
+    bool isRelative() const;
+
+    bool exists() const;
+
+    Path rootItem() const;
+    Path parentPath() const;
+    Path filename() const;
+
+    QString extension() const;
+    bool hasExtension(const QString &ext, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    void removeExtension();
+
+    bool isEqualTo(const Path &other, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    bool isBaseOf(const Path &other, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive) const;
+    Path toRelativePath(const Path &basePath) const;
+
+    QString toString() const;
+    QString toNative() const;
+
+    Path &operator=(const Path &other) = default;
+    Path &operator=(Path &&other) = default;
+    Path &operator/=(const Path &other);
+    Path &operator+=(const QString &str);
+    Path &operator+=(const char str[]);
+    Path &operator+=(const std::string &str);
+
+    static Path commonBasePath(const Path &left, const Path &right, Qt::CaseSensitivity caseSensitivity = Qt::CaseSensitive);
+
+    static Path findRootFolder(const QVector<Path> &filePaths);
+    static void stripRootFolder(QVector<Path> &filePaths);
+    static void addRootFolder(QVector<Path> &filePaths, const Path &rootFolder);
+
+private:
+    friend bool operator==(const Path &lhs, const Path &rhs);
+    friend bool operator!=(const Path &lhs, const Path &rhs);
+    friend Path operator/(const Path &lhs, const Path &rhs);
+    friend Path operator+(const Path &lhs, const QString &rhs);
+    friend Path operator+(const Path &lhs, const char rhs[]);
```

Comment:
> for function parameter it should be `const char *rhs`.

It is intention to use `const char[]` since it should be applicable to string literals written in the code only. Any other char arrays shouldn't be treated as latin1 strings.

Reply:
>Any other char arrays shouldn't be treated as latin1 strings.

Or is it possible to use `QLatin1String` as the parameter instead?
Using `char[]` or `char*` can't really guard against misuse (non latin1 strings/literals).

Reply:
> is it possible to use `QLatin1String` as the parameter instead?

No. Unless you want to go back to where we started:
```c++
Path(QLatin1String("some/predefined/path"))
```
Even in this case it's enough to have overloads with `QString` parameter.

> Using `char[]` or `char*` can't really guard against misuse (non latin1 strings/literals).

`QLatin1String` doesn't do it either. This is just a way to make the conversion easier/faster when the developer is aware that the string is latin1. In other words, the `Path` class provides some contract in the interface that it assumes `latin-1` to be the encoding of the string literal (unlike QString, which uses UTF-8 for string literals and char arrays).

Reply:
>the Path class provides some contract in the interface that it assumes latin-1 to be the encoding of the string literal 

I was hoping it can be enforced either at compile-time or run-time (under debug mode)...